File: src/pt/joaomneto/ffgbutil/adventure/AdventureFragment.java
package pt.joaomneto.ffgbutil.adventure;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import android.support.v4.app.DialogFragment;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;

public abstract class AdventureFragment extends DialogFragment {

	public abstract void refreshScreensFromResume();

	@Override
	public void onResume() {
		super.onResume();
		refreshScreensFromResume();
	}

	public void setupIncDecButton(View rootView, Integer incButtonId, Integer decButtonId, final String getMethod, final String setMethod, final Integer maxValue) {
		setupIncDecButton(rootView, incButtonId, decButtonId, null, null, getMethod, setMethod, maxValue);
	}

	public void setupIncDecButton(View rootView, Integer incButtonId, Integer decButtonId, final Adventure adv, @SuppressWarnings("rawtypes") final Class clazz, final String getMethod,
			final String setMethod, final Integer maxValue) {
		setupIncDecButton(rootView, incButtonId, decButtonId, adv, clazz, getMethod, setMethod, maxValue, null, null);
	}

	public void setupIncDecButton(View rootView, Integer incButtonId, Integer decButtonId, final Adventure adv, @SuppressWarnings("rawtypes") final Class clazz, final String getMethod,
			final String setMethod, final Integer maxValue, final Runnable incTrigger, final Runnable decTrigger) {

		Button incButton = (Button) rootView.findViewById(incButtonId);
		Button decButton = (Button) rootView.findViewById(decButtonId);

		final Object this_ = this;

		incButton.setOnClickListener(new OnClickListener() {

			@Override
			public void onClick(View v) {
				try {

					incDec(adv, clazz, getMethod, setMethod, maxValue, this_, true);

					if (incTrigger != null)
						incTrigger.run();

					refreshScreensFromResume();
				} catch (Exception e) {
					Log.e("Adventure", e.getMessage());
				}
			}

		});

		decButton.setOnClickListener(new OnClickListener() {

			@Override
			public void onClick(View v) {
				try {

					incDec(adv, clazz, getMethod, setMethod, maxValue, this_, false);

					if (decTrigger != null)
						decTrigger.run();

					refreshScreensFromResume();
				} catch (Exception e) {
					Log.e("Adventure", e.getMessage());
				}
			}

		});

	}

	@SuppressWarnings({ "unchecked", "rawtypes" })
	private void incDec(final Adventure adv, final Class clazz, final String getMethod, final String setMethod, final Integer maxValue, final Object this_, boolean increase)
			throws NoSuchMethodException, IllegalAccessException, InvocationTargetException {
		Method mGetMethod;
		Method mSetMethod;
		Object instance;

		if (adv != null) {
			mGetMethod = clazz.getMethod(getMethod);
			mSetMethod = clazz.getMethod(setMethod, int.class);
			instance = adv;
		} else {
			mGetMethod = this_.getClass().getMethod(getMethod);
			mSetMethod = this_.getClass().getMethod(setMethod, int.class);
			instance = this_;
		}

		if (increase)
			mSetMethod.invoke(instance, Math.min(((Integer) mGetMethod.invoke(instance)) + 1, maxValue));
		else
			mSetMethod.invoke(instance, Math.max(((Integer) mGetMethod.invoke(instance)) - 1, 0));
	}


//		@Override
//		public void setUserVisibleHint(boolean isVisibleToUser) {
//			super.setUserVisibleHint(isVisibleToUser);
//			if (isVisibleToUser) {
//				refreshScreensFromResume();
//			}
//			else {  }
//		}


}

File: src/main/java/pt/joaomneto/titancompanion/adventure/AdventureFragment.kt
package pt.joaomneto.titancompanion.adventure

import android.support.v4.app.DialogFragment
import android.view.View
import android.widget.Button
import kotlin.reflect.KMutableProperty0
import kotlin.reflect.KMutableProperty1

abstract class AdventureFragment : DialogFragment() {

    var baseLayout = -1

    abstract fun refreshScreen()

    override fun onResume() {
        super.onResume()
        refreshScreen()
    }

    fun setupIncDecButton(
        rootView: View,
        incButtonId: Int,
        decButtonId: Int,
        property: KMutableProperty0<Int>,
        maxValue: Int
    ) {
        val incButton = rootView.findViewById<Button>(incButtonId)
        val decButton = rootView.findViewById<Button>(decButtonId)

        incButton.setOnClickListener {
            incDec(property, maxValue, true)
            refreshScreen()
        }

        decButton.setOnClickListener {
            incDec(property, maxValue, false)
            refreshScreen()
        }
    }

//    fun <A : Adventure> setupIncDecButton(
//        rootView: View, incButtonId: Int, decButtonId: Int, adv: A, property: KMutableProperty1<A, Int>,
//        maxValue: Int, incTrigger: Runnable? = null, decTrigger: Runnable? = null
//    ) {
//
//        val incButton = rootView.findViewById<Button>(incButtonId)
//        val decButton = rootView.findViewById<Button>(decButtonId)
//
//        incButton.setOnClickListener {
//            incDec(adv, property, maxValue, true)
//            incTrigger?.run()
//            refreshScreen()
//        }
//
//        decButton.setOnClickListener {
//            incDec(adv, property, maxValue, false)
//            decTrigger?.run()
//            refreshScreen()
//        }
//    }
//
//    private fun <A : Adventure> incDec(adv: A, property: KMutableProperty1<A, Int>, maxValue: Int, increase: Boolean) {
//        if (increase)
//            property.set(adv, Math.min(property.get(adv) + 1, maxValue))
//        else
//            property.set(adv, Math.min(property.get(adv) - 1, 0))
//    }

    private fun incDec(property: KMutableProperty0<Int>, maxValue: Int, increase: Boolean) {
        if (increase)
            property.set(Math.min(property.get() + 1, maxValue))
        else
            property.set(Math.min(property.get() - 1, 0))
    }
}
